### PR TITLE
refactor addins indexer to provide extensible package resource indexer

### PR DESCRIPTION
**Pull request by kevinushey**
NOTE: not quite ready for merge; needs some testing to ensure nothing is regressed on the addins side; we might also want to tweak the overall implementation as well depending on whether we're happy with the current separation of concerns.

---

This PR attempts to refactor the indexer we use for R addins into a more general base class that should be useful for handling of project templates provided by R packages.

Most of the machinery used in the R addins indexer has been moved to a more general class, which is tentatively called the 'package provided extension' indexer (or `ppe::Indexer` for short). The general idea is that users will subclass this indexer, implementing a couple of virtual methods to control how indexing is handled, and the classes will define a method that describes how to rebuild those objects from a file provided by an R package.

### CODE REVIEW

**Comment by kevinushey** [comment]
I think this PR should be ready for review; @jjallaire, do you want to sanity check?

**Comment by jjallaire** [comment]
Yup, will do tomorrow or first thing Monday.

File: src/cpp/session/include/session/SessionPackageProvidedExtension.hpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+#include <session/SessionModuleContext.hpp>
+
+namespace rstudio {
+namespace session {
+namespace modules {
+namespace ppe {
+
+class Indexer : public boost::noncopyable
+{
+   virtual void onIndexingStarted() = 0;
```
Comment (jjallaire) [inline src/cpp/session/include/session/SessionPackageProvidedExtension.hpp]: You'd want a virtual destructor here as well

Comment (kevinushey) [inline src/cpp/session/include/session/SessionPackageProvidedExtension.hpp]: Good catch; thanks!

**Comment by jjallaire** [comment]
I think this looks good. Merge when ready!

**Pull request merged**